README.md
# MMM-NewsFeed

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/).

You can display news from different RSS feeds.

## Installing the module

Clone this repository in your MagicMirror/modules folder.

```bash
cd ./MMM-NewsFeed
npm i
```

## Using the module

To use this module, add the following configuration block to the modules array in the `config/config.js` file:
```js
var config = {
    modules: [
        {
            module: 'MMM-NewsFeed',
            config: {
                feeds: [
                			{
                				title: "bild.de",
                				url: "https://www.bild.de/rssfeeds/vw-home/vw-home-16725562,dzbildplus=true,sort=1,teaserbildmobil=false,view=rss2.bild.xml"
                			}
                		],
                showSourceTitle: true,
                showPublishDate: true,
                showImage: true,
                showDescription: true,
                truncDescription: true,
                lengthDescription: 500,
                hideLoading: false,
                reloadInterval: 5 * 60 * 1000, // every 5 minutes
                updateInterval: 30 * 1000, //every 30 seconds
                animationSpeed: 2.5 * 1000,
                maxNewsItems: 0,
                ignoreOldItems: false,
                ignoreOlderThan: 2 * 24 * 60 * 60 * 1000, // 2 days
                logFeedWarnings: false
            }
        }
    ]
}
```

## Configuration options

| Option              | Description
|-------------------- |-----------
| `feeds`             | *Required* List of RSS feeds with a title and url. If you have trouble with encoding you can specify a different one wit the encoding-attribute.
| `showSourceTitle`   | *Optional* Boolean property to show the source of the feed (e. g. 'bild.de')
| `showPublishDate`   | *Optional* Boolean property to show the puplish date of the feed item (e. g. '11 hours ago')
| `showImage`         | *Optional* Boolean property. If the feed provides an image it will be displayed
| `showDescription`   | *Optional* Boolean property to show a description of the news item
| `truncDescription`  | *Optional* Boolean property to shorten the description if too long
| `lengthDescription` | *Optional* Integer value to define how many letters the description should contain
| `hideLoading`       | *Optional* Boolean property to hide the loading label on startup 
| `reloadInterval`    | *Optional* Integer value to reload the known RSS feeds
| `updateInterval`    | *Optional* Integer value of display duration of one news item
| `animationSpeed`    | *Optional* Integer value of the fading speed between news items
| `maxNewsItems`      | *Optional* Integer value of the maximum news items to load
| `ignoreOldItems`    | *Optional* Boolean property if older news items should be displayed
| `ignoreOlderThan`   | *Optional* Integer value for maximum age from news items
| `logFeedWarnings`   | *Optional* Boolean property for logging feed warnings when consuming




CHANGELOG.md
# MMM-NewsFeed Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [1.0.0] - Unreleased

First public release


fetcher.js
/* Magic Mirror
 * Fetcher
 *
 * By Michael Teeuw http://michaelteeuw.nl
 * MIT Licensed.
 */

const FeedMe = require("feedme");
const request = require("request");
const iconv = require("iconv-lite");
const getUrls = require("get-urls");

/* Fetcher
 * Responsible for requesting an update on the set interval and broadcasting the data.
 *
 * attribute url string - URL of the news feed.
 * attribute reloadInterval number - Reload interval in milliseconds.
 * attribute logFeedWarnings boolean - Log warnings when there is an error parsing a news article.
 */

const Fetcher = function (url, reloadInterval, encoding, logFeedWarnings) {
	const self = this;
	if (reloadInterval < 1000) {
		reloadInterval = 1000;
	}

	let reloadTimer = null;
	let items = [];

	let fetchFailedCallback = function () {
	};
	let itemsReceivedCallback = function () {
	};

	/* private methods */

	/* fetchNews()
	 * Request the new items.
	 */

	const fetchNews = function () {
		clearTimeout(reloadTimer);
		reloadTimer = null;
		items = [];

		const parser = new FeedMe();

		parser.on("item", function (item) {
			const title = item.title;
			let description = item.description || item.summary || item.content || "";
			let imageSearch = item.description + item["content:encoded"] + item.summary + item.content || "";
			const pubdate = item.pubdate || item.published || item.updated || item["dc:date"];
			const url = item.url || item.link || "";

			//console.log("Title: " + title);
			//console.log("Description: " + description);
			//console.log("ImageSearch: " + imageSearch);
			//console.log("Pubdate: " + pubdate);
			//console.log("Url: " + url);
			if (title && pubdate) {
				const urlMatches = Array.from(getUrls(imageSearch));
				const qualityUrlMatches = [];
				urlMatches.filter(match => match.match(/([a-z\-_0-9\/\:\.]*\.(jpg|jpeg|png|gif))/i)).forEach(match => {
					qualityUrlMatches.push(match.replace(",w=120,", ",w=400,").replace("-thumbsmall-", "-thumb-"))
				});
				const cleanRegex = /(<([^>]+)>)/ig;
				description = description.toString().replace(cleanRegex, "");
				items.push({
					title: title,
					description: description,
					pubdate: pubdate,
					url: url,
					images: qualityUrlMatches
				});

			} else if (logFeedWarnings) {
				console.log("Can't parse feed item:");
				console.log(item);
				console.log("Title: " + title);
				console.log("Description: " + description);
				console.log("Pubdate: " + pubdate);
			}
		});

		parser.on("end", function () {
			//console.log("end parsing - " + url);
			self.broadcastItems();
			scheduleTimer();
		});

		parser.on("error", function (error) {
			fetchFailedCallback(self, error);
			scheduleTimer();
		});

		const headers = {
			"User-Agent": "Mozilla/5.0 (Node.js " + Number(process.version.match(/^v(\d+\.\d+)/)[1]) + ") MagicMirror/" + global.version + " (https://github.com/MichMich/MagicMirror/)",
			"Cache-Control": "max-age=0, no-cache, no-store, must-revalidate",
			"Pragma": "no-cache"
		}

		request({uri: url, encoding: null, headers: headers})
			.on("error", function (error) {
				fetchFailedCallback(self, error);
				scheduleTimer();
			})
			.pipe(iconv.decodeStream(encoding)).pipe(parser);

	};

	/* scheduleTimer()
	 * Schedule the timer for the next update.
	 */
	var scheduleTimer = function () {
		//console.log('Schedule update timer.');
		clearTimeout(reloadTimer);
		reloadTimer = setTimeout(function () {
			fetchNews();
		}, reloadInterval);
	};

	/* public methods */

	/* setReloadInterval()
	 * Update the reload interval, but only if we need to increase the speed.
	 *
	 * attribute interval number - Interval for the update in milliseconds.
	 */
	this.setReloadInterval = function (interval) {
		if (interval > 1000 && interval < reloadInterval) {
			reloadInterval = interval;
		}
	};

	/* startFetch()
	 * Initiate fetchNews();
	 */
	this.startFetch = function () {
		fetchNews();
	};

	/* broadcastItems()
	 * Broadcast the existing items.
	 */
	this.broadcastItems = function () {
		if (items.length <= 0) {
			//console.log('No items to broadcast yet.');
			return;
		}
		//console.log('Broadcasting ' + items.length + ' items.');
		itemsReceivedCallback(self);
	};

	this.onReceive = function (callback) {
		itemsReceivedCallback = callback;
	};

	this.onError = function (callback) {
		fetchFailedCallback = callback;
	};

	this.url = function () {
		return url;
	};

	this.items = function () {
		return items;
	};
};

module.exports = Fetcher;


node_helper.js
/* Magic Mirror
 * Node Helper: Newsfeed
 *
 * By Michael Teeuw http://michaelteeuw.nl
 * MIT Licensed.
 */

const NodeHelper = require("node_helper");
const validUrl = require("valid-url");
const Fetcher = require("./fetcher.js");

module.exports = NodeHelper.create({
	// Subclass start method.
	start: function () {
		console.log("Starting module: " + this.name);
		this.fetchers = [];
	},

	// Subclass socketNotificationReceived received.
	socketNotificationReceived: function (notification, payload) {
		if (notification === "MMM-NEWSFEED_ADD_FEED") {
			this.createFetcher(payload.feed, payload.config);
			return;
		}
	},

	/* createFetcher(feed, config)
	 * Creates a fetcher for a new feed if it doesn't exist yet.
	 * Otherwise it reuses the existing one.
	 *
	 * attribute feed object - A feed object.
	 * attribute config object - A configuration object containing reload interval in milliseconds.
	 */
	createFetcher: function (feed, config) {
		const self = this;

		const url = feed.url || "";
		const encoding = feed.encoding || "UTF-8";
		const reloadInterval = feed.reloadInterval || config.reloadInterval || 5 * 60 * 1000;

		if (!validUrl.isUri(url)) {
			self.sendSocketNotification("MMM-NEWSFEED_INCORRECT_URL", url);
			return;
		}

		let fetcher;
		if (typeof self.fetchers[url] === "undefined") {
			console.log("Create new news fetcher for url: " + url + " - Interval: " + reloadInterval);
			fetcher = new Fetcher(url, reloadInterval, encoding, config.logFeedWarnings);

			fetcher.onReceive(function (fetcher) {
				self.broadcastFeeds();
			});

			fetcher.onError(function (fetcher, error) {
				self.sendSocketNotification("MMM-NEWSFEED_FETCH_ERROR", {
					url: fetcher.url(),
					error: error
				});
			});

			self.fetchers[url] = fetcher;
		} else {
			console.log("Use existing news fetcher for url: " + url);
			fetcher = self.fetchers[url];
			fetcher.setReloadInterval(reloadInterval);
			fetcher.broadcastItems();
		}

		fetcher.startFetch();
	},

	/* broadcastFeeds()
	 * Creates an object with all feed items of the different registered feeds,
	 * and broadcasts these using sendSocketNotification.
	 */
	broadcastFeeds: function () {
		const feeds = {};
		for (let f in this.fetchers) {
			feeds[f] = this.fetchers[f].items();
		}

		this.sendSocketNotification("MMM-NEWSFEED_NEWS_ITEMS", feeds);
	}
});


MMM-NewsFeed.js
/* global Module */

/* Magic Mirror
 * Module: MMM-NewsFeed
 *
 * By Stefan Nachtrab
 * MIT Licensed.
 */

Module.register("MMM-NewsFeed", {
	// Default module config.
	defaults: {
		feeds: [
			{
				title: "bild.de",
				url: "https://www.bild.de/rssfeeds/vw-home/vw-home-16725562,dzbildplus=true,sort=1,teaserbildmobil=false,view=rss2.bild.xml"
			}
		],
		showSourceTitle: true,
		showPublishDate: true,
		showImage: true,
		showDescription: true,
		truncDescription: true,
		lengthDescription: 500,
		hideLoading: false,
		reloadInterval: 5 * 60 * 1000, // every 5 minutes
		updateInterval: 30 * 1000, // every 30 seconds
		animationSpeed: 2.5 * 1000,
		maxNewsItems: 0, // 0 for unlimited
		ignoreOldItems: false,
		ignoreOlderThan: 2 * 24 * 60 * 60 * 1000, // 2 days
		logFeedWarnings: false
	},

	requiresVersion: "2.1.0", // Required version of MagicMirror

	getScripts: function () {
		return ["moment.js"];
	},
	// Define required translations.
	getTranslations: function () {
		return false;
	},
	start: function () {
		Log.info("Starting module: " + this.name);

		// Set locale.
		moment.locale(config.language);

		this.newsItems = [];
		this.loaded = false;
		this.activeItem = 0;

		this.registerFeeds();
	},
	// Override socket notification handler.
	socketNotificationReceived: function (notification, payload) {
		if (notification === "MMM-NEWSFEED_NEWS_ITEMS") {
			Log.info("payload", payload);
			this.generateFeed(payload);

			if (!this.loaded) {
				this.scheduleUpdateInterval();
			}

			this.loaded = true;
		}
	},
	// Override dom generator.
	getDom: function () {
		const wrapper = document.createElement("div");

		if (this.activeItem >= this.newsItems.length) {
			this.activeItem = 0;
		}

		if (this.newsItems.length > 0) {
			const parentPanel = document.createElement("div");
			parentPanel.className = "parentPanel";
			if (this.config.showImage && this.newsItems[this.activeItem].images.length > 0) {
				const imagePanel = document.createElement("div");
				imagePanel.className = "imagePanel align-middle";
				const image = document.createElement("img");
				image.className = "img-fluid";
				image.width = 300;
				image.src = this.newsItems[this.activeItem].images[0];
				imagePanel.appendChild(image);
				wrapper.appendChild(imagePanel);
			}

			const textPanel = document.createElement("div");
			textPanel.className = "textPanel align-middle";
			wrapper.appendChild(textPanel);

			if (this.config.showSourceTitle || this.config.showPublishDate) {
				const sourceAndTimestamp = document.createElement("div");
				sourceAndTimestamp.className = "light small dimmed";

				if (this.config.showSourceTitle && this.newsItems[this.activeItem].sourceTitle !== "") {
					sourceAndTimestamp.innerHTML = this.newsItems[this.activeItem].sourceTitle;
				}
				if (this.config.showSourceTitle && this.newsItems[this.activeItem].sourceTitle !== "" && this.config.showPublishDate) {
					sourceAndTimestamp.innerHTML += ", ";
				}
				if (this.config.showPublishDate) {
					sourceAndTimestamp.innerHTML += moment(new Date(this.newsItems[this.activeItem].pubdate)).fromNow();
				}
				if (this.config.showSourceTitle && this.newsItems[this.activeItem].sourceTitle !== "" || this.config.showPublishDate) {
					sourceAndTimestamp.innerHTML += ":";
				}

				textPanel.appendChild(sourceAndTimestamp);
			}

			const title = document.createElement("div");
			title.className = "bright medium light";
			title.innerHTML = this.newsItems[this.activeItem].title;
			textPanel.appendChild(title);

			if (this.config.showDescription) {
				const description = document.createElement("div");
				description.className = "small light";
				const txtDesc = this.newsItems[this.activeItem].description;
				description.innerHTML = (this.config.truncDescription ? (txtDesc.length > this.config.lengthDescription ? txtDesc.substring(0, this.config.lengthDescription) + "..." : txtDesc) : txtDesc);
				textPanel.appendChild(description);
			}

			if (this.config.hideLoading) {
				this.show();
			}

		} else {
			if (this.config.hideLoading) {
				this.hide();
			} else {
				wrapper.innerHTML = this.translate("LOADING");
				wrapper.className = "small dimmed";
			}
		}

		return wrapper;
	},
	/* registerFeeds()
	 * registers the feeds to be used by the backend.
	 */
	registerFeeds: function () {
		this.config.feeds.forEach(feed => {
			this.sendSocketNotification("MMM-NEWSFEED_ADD_FEED", {
				feed: feed,
				config: this.config
			});
		});
	},
	/* generateFeed()
	 * Generate an ordered list of items for this configured module.
	 *
	 * attribute feeds object - An object with feeds returned by the node helper.
	 */
	generateFeed: function (feeds) {
		let newsItems = [];
		for (const feed in feeds) {
			Log.info("feed", feed);
			const feedItems = feeds[feed];
			if (this.subscribedToFeed(feed)) {
				for (const item of feedItems) {
					item.sourceTitle = this.titleForFeed(feed);
					if (!(this.config.ignoreOldItems && ((Date.now() - new Date(item.pubdate)) > this.config.ignoreOlderThan))) {
						newsItems.push(item);
					}
				}
			}
		}
		newsItems.sort(function (a, b) {
			const dateA = new Date(a.pubdate);
			const dateB = new Date(b.pubdate);
			return dateB - dateA;
		});
		if (this.config.maxNewsItems > 0) {
			newsItems = newsItems.slice(0, this.config.maxNewsItems);
		}

		this.newsItems = newsItems;
	},
	/* subscribedToFeed(feedUrl)
		 * Check if this module is configured to show this feed.
		 *
		 * attribute feedUrl string - Url of the feed to check.
		 *
		 * returns bool
		 */
	subscribedToFeed: function (feedUrl) {
		for (const feed of this.config.feeds) {
			if (feed.url === feedUrl) {
				return true;
			}
		}
		return false;
	},
	/* titleForFeed(feedUrl)
		 * Returns title for a specific feed Url.
		 *
		 * attribute feedUrl string - Url of the feed to check.
		 *
		 * returns string
		 */
	titleForFeed: function (feedUrl) {
		for (const feed of this.config.feeds) {
			if (feed.url === feedUrl) {
				return feed.title || "";
			}
		}
		return "";
	},
	/* scheduleUpdateInterval()
	 * Schedule visual update.
	 */
	scheduleUpdateInterval: function () {
		const self = this;

		self.updateDom(self.config.animationSpeed);

		setInterval(function () {
			self.activeItem++;
			self.updateDom(self.config.animationSpeed);
		}, this.config.updateInterval);
	},
	getStyles: function () {
		return [
			"MMM-NewsFeed.css",
		];
	}
});
